design/ceph/update-osds-in-parallel.md

```diff
@@ -0,0 +1,174 @@
+> parallel without regard to failure domain? For example, try to update some number (or percentage)
+> of OSDs at once in any failure domain. This might be a good default strategy ("best-effort"?).
+
+### Ability to set a maximumum number of OSDs to be updated simultaneously
+In the worst (but unlikely) case, all OSDs which are updated for a given parallel update operation
+might fail to come back online after they are updated. Users may wish to limit the number of OSDs
+updated in parallel in order to avoid too many OSDs failing in this way.
+
+### Cluster growth takes precedence over updates
+Adding new OSDs to a cluster should occur before updating currently-running OSDs. This allows users
+to make use of newly added storage as quickly as possible, which they may need for critical
+applications using the underlying Ceph storage. In some degraded cases, adding new storage may be
+necessary in order to allow currently-running Ceph OSDs to be updated without experiencing storage
+cluster downtime.
+
+This prioritization might delay updates significantly since adding OSDs not only adds capacity to 
+the Ceph cluster but also necessitates data rebalancing. Rebalancing generates data movement which
+needs to settle for updates to be able to proceed.
+
+### OSD updates should not starve other resources of updates
+For Ceph cluster with huge numbers of OSDs, Rook's process to update OSDs should not starve other
+resources out of the opportunity to get configuration updates.
+
+
+## Technical implementation details
+
+### Workflow
+Use the `ceph osd ok-to-stop <id>...` command to determine if it is safe to update multiple OSDs at
+once.
+
+1. Build an "update queue" of OSDs which need updated.
+2. Query `ok-to-stop` for a single OSD from the update queue. This represents the CRUSH hierarchy
+   bucket at the"osd" (or "device") level.
+3. If the previous operation reports that it safe to update, batch query `ok-to-stop` for all OSDs
+   that fall under the CRUSH bucket one level up from the current level.
+4. Repeat step 3 moving up the CRUSH hierarchy until one of the following two conditions:
+    1. The number of OSDs in the batch query is greater or equal to the maximum number of OSDs the
+       user allows to be updated in parallel, OR
+    2. It is no longer `ok-to-stop` all OSDs in the CRUSH bucket.
+5. Update OSD Deployments in parallel for the last CRUSH bucket where it was `ok-to-stop` the OSDs. 
+    - If there are more OSDs in the CRUSH bucket than allowed by the user, update OSDs in the CRUSH
+      bucket in groups as large as the user configuration allows until all OSDs in the CRUSH bucket 
+      are updated.
+    - When an OSD is updated successfully, remove it from the update queue.
+7. Return to step 2 until all OSDs are updated.
+
+### How to make sure Ceph cluster updates are not starved by OSD updates
+Because [cluster growth takes precedence over updates](#cluster-growth-takes-precedence-over-updates),
```

Comment:
Using the same checkpoint strategy introduced in #6693 may be sufficient to cancel a reconcile in progress so we don't block too long.